File: NPuzzleAStar/npuzzle/npuzzle/AstarWeb/nPuzzleWeb/p5/nPuzzle/scramblerClass.js
class Scrambler
{
	constructor(amount)
	{
		this.all	 = ["UP","DOWN","RIGHT","LEFT"];
		this.noUp	 = ["RIGHT","LEFT"];
		this.noDown	 = ["RIGHT","LEFT"];
		this.noRight = ["UP","DOWN"];
		this.noLeft	 = ["UP","DOWN"];

		this.lastMove = null;

		this.max = amount;		// variable indiquant le nombre maximum de mouvement que le npuzzle va executer pour le scrambler
		this.isOn = true; // boolean indiquant si le npuzzle à été au moins scramble une fois.
		this.countMove = 0;
	}
}

Scrambler.prototype.applyMove = function(nextMove)
{
	if (nextMove == this.lastMove)
		return false;
	if (nextMove == "UP")
		return npuzzle.Up();
	else if (nextMove == "DOWN")
		return npuzzle.Down();
	else if (nextMove == "RIGHT")
		return npuzzle.Right();
	else if (nextMove == "LEFT")
		return npuzzle.Left();
}

Scrambler.prototype.oneRandomMove = function()
{
	let m = null;
	if (this.isOn == false)
		return;
	this.countMove = (this.countMove + 1) % this.max;
	while (1)
	{
		if (this.lastMove == "UP")
			m = this.noUp[getRandomInt(2)];
		else if (this.lastMove == "RIGHT")
			m = this.noRight[getRandomInt(2)];
		else if (this.lastMove == "DOWN")
			m = this.noDown[getRandomInt(2)];
		else if (this.lastMove == "LEFT")
			m = this.noLeft[getRandomInt(2)];
		else if (this.lastMove == null)
			m = this.all[getRandomInt(4)];
		let ret = this.applyMove(m);
		if (ret == true)
			break ;
	}
	if (this.countMove == 0)
		this.isOn = false;
//	console.log(m);
	this.lastMove = m;
}

File: NPuzzleAStar/npuzzle/npuzzle/src/Scrambler.class.cpp
#include "../includes/Scrambler.class.hpp"

Scrambler::Scrambler(void) : lastMove("")
{
  srand(time(nullptr));
}

Scrambler::~Scrambler(void) {}


bool Scrambler::applyMove(std::string move, nPuzzle &puzzle)
{
//  cout << "move : " << move << endl;
  if (move == lastMove)
    return false;
  if (move == "UP")
    return puzzle.Up();
  else if (move == "RIGHT")
    return puzzle.Right();
  else if (move == "DOWN")
    return puzzle.Down();
  else if (move == "LEFT")
    return puzzle.Left();
  return false;
}

void Scrambler::scramble(short n, nPuzzle &puzzle)
{
  std::string m = _moves[rand() % _moves.size()];
  short       v = 0;
  short counterFail = 0;
  short limitFail = 500;
  for (short i = 0; i < n; ++i)
  {
      counterFail = 0;
      while (true)
      {
        if (lastMove == "UP")
          m = _noUp[rand() % _noUp.size()];
        else if (lastMove == "RIGHT")
          m = _noRight[rand() % _noRight.size()];
        else if (lastMove == "DOWN")
          m = _noDown[rand() % _noDown.size()];
        else if (lastMove == "LEFT")
          m = _noLeft[rand() % _noLeft.size()];
          if (applyMove(m, puzzle) == true)
            break;
          if (counterFail++ == limitFail)
            m = _moves[rand() % _moves.size()];
      }
      lastMove = m;
  }
}
